extensions/in-memory/assetindex-memory/src/main/java/org/eclipse/dataspaceconnector/metadata/memory/InMemoryAssetIndex.java

```diff
@@ -15,28 +15,35 @@
 package org.eclipse.dataspaceconnector.metadata.memory;
 
 import org.eclipse.dataspaceconnector.spi.asset.AssetIndex;
+import org.eclipse.dataspaceconnector.spi.asset.AssetIndexLoader;
 import org.eclipse.dataspaceconnector.spi.asset.AssetSelectorExpression;
-import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.asset.DataAddressResolver;
 import org.eclipse.dataspaceconnector.spi.types.domain.asset.Asset;
 import org.eclipse.dataspaceconnector.spi.types.domain.transfer.DataAddress;
 
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
- * An ephemeral asset index.
+ * An ephemeral asset index, that is also a DataAddressResolver and an AssetIndexLoader
  */
-public class InMemoryAssetIndex implements AssetIndex {
+public class InMemoryAssetIndex implements AssetIndex, DataAddressResolver, AssetIndexLoader {
     private final Map<String, Asset> cache = new ConcurrentHashMap<>();
+    private final Map<String, DataAddress> dataAddresses = new ConcurrentHashMap<>();
     private final CriterionToPredicateConverter predicateFactory;
+    private final ReentrantLock lock;
```

Comment:
why do we need this lock in addition to ConcurrentHashMaps? Just to ensure both Maps stay consistent?
If so maybe a ReentrantReadWriteLock would be better... but just a minor-minor improvement for an extension that shouldn't be used in production...

Reply:
Are you after the supposed added performance since we're expecting much more `read` operations than `write` operations?

Reply:
yes. but more than that I wanted to make ensure I understood why we need the lock at all? Is it for consistency over those two maps?

Reply:
If so my next question would be: why is DataAddress not just an attribute at Asset? From a DDD perspective: does a DataAddress has it's own identity? Don't think so - but maybe I'm wrong...

Reply:
Actually looking from a DDD perspective the `Asset` is an entity, it has an identity, it has an history and, why not, their attributes could be modified without changing its identity. 
Seems like `DataAddress` in this context is more of a value object, because it is just related to the `Asset`, and in fact it does not own an identity so I agree with your reasoning @MoritzKeppler .
Don't remember why we extracted `DataAddress` from `Asset`...

Reply:
yes exactly. My intention was to mimic the behavior of a database where inserting values into two tables does not happen at the exact same time, but rather one-after-the-other. A transaction takes care of consistency.

Reply:
I would not model the `DataAddress` as a property of the `Asset`, because the DataAddress will probably private information - we wouldn't want to expose where our data is stored. Thus, if the DataAddress is a property on the Asset, we would have to make prevent exposing it during serialization, which I find dangerous.

Reply:
What we offer to outside is e.g. an IDS Resource or something similar. In my eyes it would be an antipattern to provide the internal model 1:1 to the world outside.

Reply:
But a DataAdress can also be used in the information passed to others: it that case it would be the address under which the other participant can retrieve the data.
As soon as a consumer retrieved an asset it will be added to an internal catalog (as it now needs to be managed) and will get a DataAddress pointing to the local copy of the asset.

Reply:
>  But a DataAdress can also be used in the information passed to others: it that case it would be the address under which the other participant can retrieve the data.

that would imply giving consumers direct access to, say, a database, right? Because I think the DataAddress would contain database name, etc...

> As soon as a consumer retrieved an ...

sure, but then there will also be a different instance of the Asset within the consumer's realm, even if that may contain the same metadata.

Reply:
started a discussion, maybe we can copy over the different views.
https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/discussions/214